File: LeetCode/75. Sort Colors.py
# https://leetcode.com/problems/sort-colors/description/

from test import test


class Solution:
    def sortColors(self, nums: list[int]) -> None:
        last_left = -1
        first_right = len(nums)

        index = 0

        while index < len(nums):
            value = nums[index]

            if value == 0:
                nums[index], nums[last_left + 1] = nums[last_left + 1], nums[index]
                last_left += 1
                index += 1
                continue

            if value == 2 and index < first_right:
                nums[index], nums[first_right - 1] = nums[first_right - 1], nums[index]
                first_right -= 1
                continue

            index += 1


solution = Solution()


test(
    solution.sortColors,
    [
        {
            "input": [[2, 0, 2, 1, 1, 0]],
            "expected": [0, 0, 1, 1, 2, 2],
        },
        {
            "input": [[2, 0, 1]],
            "expected": [0, 1, 2],
        },
        {
            "input": [[2, 0, 1]],
            "expected": [0, 1, 2],
        },
    ],
)

File: LeetCode/75. Sort Colors.js
// https://leetcode.com/problems/sort-colors/description/
const { test } = require('../test');

const sortColors = (nums1) => {
  const nums = [...nums1];

  let lastZero = -1;
  let firstTwo = nums.length;

  for (let index = 0; index < nums.length; index++) {
    const current = nums[index];

    if (current === 0) {
      [nums[index], nums[lastZero + 1]] = [nums[lastZero + 1], nums[index]];
      lastZero++;
    }
    if (current === 2 && index < firstTwo) {
      [nums[index], nums[firstTwo - 1]] = [nums[firstTwo - 1], nums[index]];
      firstTwo--;
      index--;
    }
  }

  return nums;
};

test(sortColors, [
  {
    input: [[2, 0, 2, 1, 1, 0]],
    expected: [0, 0, 1, 1, 2, 2],
  },
  {
    input: [[2, 0, 1]],
    expected: [0, 1, 2],
  },
]);
